Paper: HOOMD-blue version 3.0 A Modern, Extensible, Flexible, Object-Oriented API for Molecular Simulations

b-butler opened the pull request:
The draft proceedings paper for submission 213.

If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

b-butler committed:
Potential paper title with authors

Used to test compiling the paper to pdf

b-butler committed:
Update 2020-05-04

b-butler committed:
Add core class diagram

b-butler committed:
Update 2020-05-08

b-butler committed:
Update 2020-05-09

b-butler committed:
Update 2020-05-10

b-butler committed:
Update 2020-05-11

b-butler committed:
Fix figure label

b-butler committed:
Update 2020-05-07

b-butler committed:
Update 2020-05-18

b-butler committed:
Updated core-object graph

b-butler committed:
Update 2020-05-21

b-butler committed:
Update 2020-05-22

vyasr committed:
Update paper.

b-butler committed:
Update 2020-05-27

b-butler committed:
Update 2020-05-28

b-butler committed:
Add paragraph explaining MoSDeF

b-butler committed:
Improve upon MoSDeF paragraph

b-butler committed:
Fix citation

b-butler committed:
Update Acknowledgements

b-butler committed:
Update code to reflect namespace

b-butler committed:
Make figure background transparent

b-butler committed:
Update custom action example

reid-a commented:
Greetings! I'm Andrew Reid, I'll be reviewing this. I had a bit of a slow start, missed an early e-mail, but I'm back on track now.

b-butler commented:
Awesome @reid-a! I look forward to working with you on improving the paper.

akhalighi commented:
Hi @b-butler and team! My name is Amir Khalighi and I'll also be reviewing your Scipy submission - You have done a great job and I feel excited to learn about your work in more detail.

b-butler commented:
Hello, @akhalighi, happy to have your thoughts and suggestions on the paper.

reid-a commented:
So I have some preliminary comments.

First of all, the paper is generally very good, I particularly like the way that the OO structure of the code is used in support of extensibility -- this can be challenging to get right, in terms of balancing "legibility" of the API against the actual object structure, and it looks like you've done a good job of exposing powerful objects without requiring users to understand too much arcane knowledge about the workings of the code.

My minor comments are small details, mostly.

In the 1st column of page 3, there are a couple of items -- you say there that you're mocking the CuPy interface, but if I understand correctly, CuPy makes a pretty strong effort to mock the Numpy interface.  Would it be perhaps clearer to your audience to just say that the interface is Numpy-like? If there is a special role played by CuPy in the interface semantics, I missed it.

Also, you mention you're using zero-copy, and that it's rank-local, as of course it must be by the nature of the zero-copy operation -- although, actually, it really just needs to be node-local, and in some MPI implementations, it's possible to have multiple ranks per node. But, that's not my issue here, it's that I didn't catch if there is a straightforward mechanism for detecting whether the data one wishes to access is node-local or not -- is this obvious from context, or do you mostly just not access data from other ranks outside of the MPI communicator framework? Or is something else going on?

For both of these, but for zero-copy in particular, does this imply a minimum version requirement for the underlying CUDA implementation? I am far from a CUDA expert, maybe any reasonably sane CUDA will have zero-copy, so this doesn't need to be mentioned?

I also have a small wording quibble with the first column of page 5, where you say that "we can provide through pybind11 a way to subclass Trigger in python...", which had me looking at the code to try to figure out what additional requirements the pybind11 wrappings impose on the subclassing operation. There don't seem to be any, the example is a completely vanilla Python subclassing operation, as far as I can tell. This is maybe the intended meaning of the sentence, i.e. that pybind11 wrappers allow perfectly ordinary subclassing to work? This is a difference  from e.g. SWIG, where inheritance from wrapped C++ classes can be tricky. If so, my misreading is anecdotal evidence that it could maybe be clearer.

I am in general cautious about putting actual code in manuscripts, but here it seems to work reasonably well, the examples seem clear and motivated, and well-connected to the discussion. It does give the manuscript a bit of an intimidating appearance, and I'm honestly not sure if you can usefully reduce the number of coded examples, but I thought I'd at least open that conversation.

As I say, generally speaking, this is a very good manuscript, it could proceed OK without revisions, but there are a few minor opportunities for improvement, in my initial assessment.

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+
+In the development of these tools, the requirements for the software to enable good science became
+more obvious. Having computational research that is transferable, reproducible, usable (by others),
+and extensible or TRUE :cite:`summers.etal2020` is necessary for fully realizing the potential of
```
Maybe capitalize the letters of the acronym TRUE.

b-butler replied:
That is a good idea.

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+users explicit and complete control over all aspects of simulation configuration. Where possible,
+the new version also provides performant, Pythonic interfaces to data stored by the C++ back end.
+Over the next few sections, we will use examples of HOOMD-blue's version 3.0 API (which is still in
+development at the time of writing) to highlight the improved extensibility, flexibilty, and ease of
```
flexibilty -> flexibility

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+The final of the three classes, :code:`Operations`, holds the different *operations* that will act
+on the simulation state. Broadly these consist of 3 categories: updaters, which modify simulation
+state; analyzers, which observe system state; and tuners, which tune the hyperparameters of other
+operations for performance. Although updaters and analzyers existed in version 2.x (tuners are a
```
analzyers -> analyzers

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+    # sets for 'B', 'C', and 'D'
+    sphere.shape[['B', 'C', 'D']] = {'diameter': 0.5}
+
+The specification defined above sets defaults for :code:`ignore_statistics` and :code:`orientatble`
```
orientatble -> orienttable

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+our logging infrastructure. The :code:`Loggable` metaclass of :code:`_Operation` allows all
+subclasses to expose their loggable quantities by marking Python properties or methods to query.
+
+The actual task of logging data is acomplished by the :code:`Logger` class, which provides an
```
acomplished -> accomplished

ocefpaf commented on papers/brandon_butler/paper.rst:
```diff
@@ -0,0 +1,715 @@
+do not have to be run on a periodic timescale; they just need some indication of when to run. In
+other words, the operations needed to be *triggered*. The :code:`Trigger` class encapsulates this
+concept providing a uniform way of specifying when an object should run without limiting options.
+:code:`Trigger` objects are essentially functors that returns a Boolean value when called.  Each
```
Not sure what `functors` are.

b-butler replied:
`functors` can have a few different meanings depending on whether you are talking about the category theory definition (the sense used in functional programing) or an object with state that acts like a function (i.e. is callable). The latter is the sense I meant it in and is the typical sense in object-oriented programming.

Function-object: https://en.wikipedia.org/wiki/Function_object
Functional Programming: https://en.wikipedia.org/wiki/Functor_(functional_programming)

However, if this is confusing I could just use the word function and use the phrase `:code:Trigger objects are like functions`.

ocefpaf replied:
I like that but you don't need to change on my account if that term is familiar to you and your audience.

b-butler replied:
I think either is just as informative, and I would want to be as inclusive as possible in my wording. Sometimes jargon is necessary, but if I can get the message across without it or put the jargon in an aside as it is now, I think that is better. Also, when you get familiar with a term you forget that it is jargon until someone points it out, so the comment was helpful.

ocefpaf commented:
Hi @b-butler and team! My name is Filipe Fernandes and I'm also reviewing this document. First of all I would like you to know that I'm mostly evaluating the text per se and not the since b/c I'm a little bit outside of my comfort zone.

I made a few suggestion on possible typos, but b/c there is a lot of "code-text" some of those words may be what you originally indented them to be.

Last but not least, some of the code blocks are short and go well with the text, but others are quite big and break the readers attention. If possible I would reduce them by hiding some boilerplate code and/or moving them to an appendix.

b-butler committed:
Add citations for mbuild and foyer

b-butler committed:
Capitalize the components of TRUE simulations

b-butler committed:
Fix spelling errors

b-butler commented:
Thanks @ocefpaf and @reid-a for your suggestions thus far.

b-butler commented:
@reid-a Thanks for the great comments.

Your thought is right, CuPy heavily mocks NumPy's API, but does not completely implement all its features. The list can be found here if you are interested https://docs-cupy.chainer.org/en/stable/reference/comparison.html. I was trying to emphasis that we only wrap around the functionality existing in CuPy which is not a full NumPy interface. I see your point though. Mentioning that we provide a NumPy-like interface that utilizes the functionality of CuPy on the GPU may be more clear to readers.

The rank-local nature of the data mimics the domain decomposition used to partition the box into local boxes for each rank. We only expose this data in the local snapshots as this is the only data that is local to the process(MPI rank). Data from a local snapshot then will always be local to the processor it is currently scheduled on (CPU) or device it is running on (GPU). I am familiar with the ability to oversubscribe a processor to more than one ranks, but from my understanding this does not change the nature of process memory access restrictions. If you are referring to process shared memory, we do not use that for state data.

For zero-copy access on the GPU, the problem would not be with the CUDA version but rather the Python libraries that support the `__cuda_array_interface__` that numba created. We use version 2 the latest, although we could use version 1 just by changing the version we advertise in the interface. As a point of reference numba requires CUDA toolkit v8 in there latest release 0.50. Given the way they document support of the protocol finding minimum versions for each package supporting the protocol would be difficult. In general this should work several versions back for the major packages supporting the protocol. For the rank local aspect, there is no problem with support.

The trigger subclassing example is indeed using pybind11 to make a vanilla subclassing work between C++ and Python classes. I will reword the section to improve the clarity.

With regard to the large amount of code, one thing I could do is move the larger examples to an appendix since they just highlight more substantial use of v3 features. I am definitely open to other suggestions here.

Edit: fixed some grammar

b-butler commented:
@ocefpaf Thanks for your review.

I appreciate the spelling catches, with all the code, I must have missed them in my spell checker.

With respect to code blocks, I mentioned in my previous comment, moving the Larger Examples to an Appendix. Are there other blocks that you see as disruptive? I don't want to disrupt the flow of the paper, but I do want to emphasis the extensibility where possible. Hiding boilerplate is definitely an option though. What is your opinion of removing import statements? I was trying to make sure all code examples worked, but if removing some lines helps readability then I am on board. If you have other idea regarding the removal or editing of code-blocks I would be happy to hear them.
Edit: grammar

ocefpaf commented:
> I appreciate the spelling catches, with all the code, I must have missed them in my spell checker.

BTW, in case you are curious, I used [codespell](https://github.com/codespell-project/codespell) for that.

> With respect to code blocks, I mentioned in my previous comment, moving the Larger Examples to an Appendix. Are there other blocks that you see as disruptive. I don't want to disrupt the flow of the paper, but I do want to emphasis the extensibility where possible. Hiding boilerplate is definitely an option though. What is your opinion of removing import statements? I was trying to make sure all code examples worked, but if removing some lines helps readability then I am on board. If you have other idea regarding the removal or editing of code-blocks I would be happy to hear them.

I have to confess that I'm on the fence. Your point is perfect valid, copy-n-pasteable examples are preferred. Maybe you can do something like "the-important-code-block-here" and a link to the appendix with the full example. But I don't want to cause a lot of churn in your paper, only do that if you believe it is easy enough and will improve the readability. Otherwise leave as-id.

PS: @deniederhut maybe we could add codespell GH action as a CI service for the proceedings PRs. If you are interested we can talk about that in an issue.

b-butler committed:
Refactor explanation of what triggers are

b-butler committed:
Update trigger subclassing discussion

mwerwee commented:
@b-butler 

I'm also reviewing this paper. My apologies for being a bit late to the game. Will work through the paper in the next few days and give feedback.

Markus

b-butler commented:
@mwerwee sounds good!

deniederhut commented:
> maybe we could add codespell GH action as a CI service for the proceedings PRs. If you are interested we can talk about that in an issue.

Absolutely!

akhalighi commented:
This is a great work and I wanted to congratulate @b-butler and the team on the work you have done to implement the new API for HOOMD-Blue. I think it is going to facilitate user adoption and benefit many in the scientific Python community. Please see below my old-school review of your manuscript:

**REVIEWER COMMENTS:**
The submission by Butler et al. gives an overview of the ongoing development efforts on the HOOMD-blue package for Version 3.0 release as well as the rationale behind some of their major design decisions; HOOMD-blue is a molecular dynamics (MD) and Monte Carlo (MC) simulation engine focused on speed, scalability, flexibility that was first released in 2008. The first version of HOOMD-blue allowed a limited Python scripting capability that required the scripts to be run through HOOMD-blue's modified interpreter. Version 2.0 improved HOOMD-blue's accessibility through Python by allowing the use of the package in ordinary Python scripts and it also supported running largely independent simulations in a Python single script. However, the Python API was only minimally modified compared to Version 1.0, and the newly introduced `SimulationContext` object for state management remained largely unknown to the users. The planned version 3.0 release of HOOMD-blue will be introducing a revamped transparent Pythonic interface comparable to the popular API conventions of the Scipy ecosystem.

In this article, the authors review the development of HOOMD-blue's new Python API and explain how the new API will improve the extensibility, flexibility, and usability of the package. Their overview starts with a minimal working example of the general API design. The new API introduces the `Simulation` object model which holds all the information and methods needed to assemble, configure, and run and MD or MC simulation. The logic inside the `Simulation` object models is organized in 3 separate classes as follows:

1. `Device` class which specifies what type of computational processor (CPU or GPU) will be used for the simulation.
2. `State` class which contains the system's data (e.g. particle positions, orientations, velocities, the system box)
3. `Operations` class which holds the updating, analyzing, tuning data that will act on the state to progress the simulation

The manuscript briefly explains the logic for this architecture, how it can be used efficiently for setting up simulations, and then moves on to review in more detail some of the features as well as use-case examples of the capabilities that the new API provides (like user customization). I think the manuscript is well-written and I don't have any major comments on that. Just a few minor suggested edits:

- Page 2: flexibilty -> flexibility
- Page 2: Code section: it looks like maybe `nlist = md.nlist.Cell()` should be changed to `nlist = hoomd.md.nlist.Cell()`
- Similarly, `lj = md.pair.LJ(nlist, r_cut=2.5)` -> `lj = hoomd.md.pair.LJ(nlist, r_cut=2.5)`. 
- Page 2: hows -> how
- Page 3: analzyers  -> analyzers
- Page 3: distinct -> distinction
- Page 4: acomplished -> accomplished

reid-a commented:
The changes on p. 5 improve the context of the sub-classing opportunity through pybind11, but there's a grammar issue -- the new wording is "we use pbyind11 to easily allow subclass the Trigger class...".  This should be "easily allow subclassing", or possibly "easily allow users to subclass".

On re-reading the first column of page 3, I think your explanation is good, but it would still be a useful addition to be clear about what versions of dependencies you are expecting. I get that this is a hard problem, but what you said in your response to me is very good, just a parenthetical that you require version 2 of `__cuda_array_interface` would help. Likewise, a word or two about zero-copy being related to domain decomposition might help -- again, I'd consider this optional, I'm torn because on the one hand, it is kind of obvious, but on the other hand, I can imagine someone reading this quickly (as I initially did) and trying it to get at "halo" data for some custom operation, and getting tripped up. It's also in tension with the abstract and intro, which emphasized (rightly) the scalability of HOOMD generally across many nodes and GPUs.  Probably it's excessive pedantry to get into multiple ranks in the same memory image, and to be honest, I kind of regret mentioning that.

I think an appendix for the code blocks of the larger-scale examples is a great idea, it would help with making the manuscript less intimidating, and possibly allow for a more detailed narrative and annotations in that section -- I endorse that suggestion, hopefully it's not too much work.

I'm also very impressed by my co-reviewers' typo-spotting powers, I missed all of those!

deniederhut commented:
Hey @b-butler ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

b-butler committed:
Add orcids.

b-butler committed:
Merge branch '2020' into hoomd-v3

b-butler commented:
@deniederhut thank you for informing me. @stargaser this is a great feature, thanks!

b-butler committed:
Fix orcids

b-butler committed:
Fix code example

b-butler committed:
Reword local snapshot API description

b-butler committed:
Fix typo in trigger subclassing

b-butler committed:
Reformat text file

b-butler committed:
Refactor custom action code block

b-butler committed:
Refactor some smaller code examples

b-butler committed:
Move Larger examples to Appendix

b-butler committed:
Fix grant number

b-butler committed:
Updat code comment

b-butler commented:
@akhalighi Thank you for your review and corrections. I believe I have addressed all the corrections.

b-butler commented:
@ocefpaf codespell looks awesome.

b-butler commented:
@reid-a I have reworded the section involving the `__cuda_array_interface__`, and mention more explicitly the relationship between zero-copy and domain decomposition.

b-butler commented:
I have moved the Larger Examples code blocks into an Appendix instead. I am willing to move more code-blocks there with explanations to reduce the code in the main paper, but I am not sure which if any of the others can be moved without forcing a user to move back and forth from the appendix to the main paper. For an example, the code which shows examples of user customization could be moved since they are larger in nature, but a main point of these sections is to show this ability to customize.

reid-a commented:
This is looking very good, I think to the extent my issues are remediable, they've been remedied.

The code block remains an issue, but there's a balance to be struck between, on the one hand, including sufficiently many sufficiently useful examples, and on the other hand, maintaining a good readable flow and not frightening away readers with big blocks of code.

I think the current version does this reasonably well, the in-line examples are relevant to the narrative and flow OK, and larger ones are segregated to the appendix. Of course improvements are probably possible, but at this point I don't have any specific suggestions, and I think you're probably in to the regime of diminishing returns on additional effort.

mwerwee commented:
@b-butler 

I've looked through the paper as it stands today, it looks good - I really appreciate the amount of effort your group has made with this code. 

The other reviewers have picked up and suggested aspects I would have commented on already, so there is nothing specific from me at this stage.

Looking forward to the conference!

Best,
Markus

deniederhut commented:
Hi @akhalighi and @ocefpaf ! Have the recent changes addressed your reviews?

ocefpaf commented:
> Hi @akhalighi and @ocefpaf ! Have the recent changes addressed your reviews?

Yep. This is good to go IMO.

akhalighi commented:
> 
> 
> Hi @akhalighi and @ocefpaf ! Have the recent changes addressed your reviews?

They did - This should be good to go!

deniederhut commented on papers/brandon_butler/references.bib:
```diff
@@ -0,0 +1,772 @@
+  number = {16}
+}
+
+@software{collette2008,
```
Instead of the github URL, would you mind citing Andrew's book? https://www.oreilly.com/library/view/python-and-hdf5/9781491944981/

b-butler replied:
No I do not. Thanks for pointing that out to me.

Edit: I noticed that we do not currently mention h5py in the paper. I will still change the reference, but it is not used in the paper itself.

deniederhut replied:
If it's not referenced from the paper, the better solution might be to remove the bibtex entry.

b-butler committed:
Use book citation for h5py instead of GitHub repo

b-butler committed:
Remove h5py bibtex entry

govarguz commented:
Dear @b-butler, 
I want to give some last shapes on the review process of your Ms. The functionalities added from an object oriented viewpoint, like the ones described in Figure 2, among the other examples show indeed the flexibility of the package. I have do a couple of small questions/feedback:

1. Does version 3.0 correspond to the github repository of HooMD-blue? Could you cite the corresponding repository within the article.

2.  In the Introduction part a couple of MD packages are mentioned, like Lammps and Gromacs. Whereby, as far as I know they also count with some kind of python APIs. Can you check this information and perhaps highlight the differences?

3. Also within the introduction in paragraph 2, I see you mention some Python/C++ bindings of e.g. OpenMM. According to the release time I would also cite E++ as it also offers a  Python and C++ interphase (https://doi.org/10.1016/j.cpc.2018.12.017 ) I think even part of such development was in a scipy conf proceedings a while ago (https://conference.scipy.org/proceedings/SciPy2009/paper_7/).

Finally, I would also emphasize the convergence/tendence that many MD packages are currently going through to have a python API. Perhaps, highlighting also the need of e.g. tutorials based on Jupyter NB  to round-up your conclusions.

b-butler committed:
Update conclusion

b-butler committed:
Add license and GitHub link to HOOMD-blue

b-butler committed:
Update paragraph on other MD/MC engines

b-butler committed:
Change sentence introducing MD/MC

b-butler committed:
Remove unused citation

b-butler commented:
@govarguz 
1. I added a link to HOOMD-blue's GitHub repository in the paper. I think that should be enough to show where users can find the code including the branches currently pushing version 3 forward.
1. I also mention that packages that have traditionally had other interfaces are adding Python bindings.
1. I added ESPResSo to the list of MD engines.

As for the conclusion, I make a statement about Python in the molecular simulation field, but I didn't feel that mentioning tutorials or jupyter notebooks fit with the rest of the paper.

govarguz commented:
@b-butler, looks good to me, except for reference of point 3 regarding the python to c++ interface is not linked to bibtex. This is the correct reference (https://doi.org/10.1016/j.cpc.2018.12.017 ), which I cannot find in the Ms, check your .bib file.
Best,
Horacio.

b-butler committed:
Add second ESPResSo citation

b-butler commented:
Good catch, if you look at the rst you will notice that I have added the citation, but I forgot to update the bib file.
